keps/sig-cluster-lifecycle/20190122-Certificates-copy-for-kubeadm-join--control-plane.md

```diff
@@ -0,0 +1,330 @@
+
+## Release Signoff Checklist
+
+- [x] k/enhancements [issue 357](https://github.com/kubernetes/enhancements/issues/357) in release
+      milestone and linked to KEP
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [x] Design details are appropriately documented
+- [x] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [x] Graduation criteria is in place
+- [x] "Implementation History" section is up-to-date for milestone
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG
+      meetings, relevant PRs/issues, release notes
+
+## Summary
+
+Automatic certificates copy makes easier to create HA clusters with the kubeadm tool using exactly
+the same `kubeadm init` and `kubeadm join` commands the users are familiar with.
+
+## Motivation
+
+As confirmed by the recent [kubeadm survey](https://drive.google.com/file/d/1eN9sGsdXWpurmplbEVn9UX5NiseQqlzO/view?usp=sharing),
+support for high availability cluster is one of the most requested features for kubeadm.
+
+A lot of effort was already done in kubeadm for achieving this goal, among them the redesign
+of the kubeadm config file and its graduation to beta and the implementation of the
+[`kubeadm join --control-plane workflow (KEP0015)`](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/0015-kubeadm-join-control-plane.md),
+but the solution currently in place stills requires the manual copy of cluster certificates from
+the bootstrap control-plane node to secondary control-plane nodes.
+
+This KEP introduces automatic certificates copy, eliminating the manual operation described
+above and completing the kubeadm solution for creating HA clusters.
+
+### Goals
+
+- Usability: use exactly the same `kubeadm init` and `kubeadm join` commands the users are
+  familiar with.
+- Security: provide a solution that enables the secure copy of cluster certificates between
+  control-plane nodes
+
+## Proposal
+
+### Constraints
+
+The solution described in this proposal is deeply influenced by following constraints:
+
+1. Kubeadm can execute actions only on the machine where it is running e.g. it is not
+   possible to execute actions on other nodes.
+2. `kubeadm init` and `kubeadm join` are separated actions (executed on separated machines/at
+   different times).
+3. During the join workflow, kubeadm can access the cluster only using identities with
+   limited grants, namely `system:unauthenticated` or `system:node-bootstrapper`.
+
+### Key element of the proposal
+
+At a high level, the proposed solution can be summarized by the following workflow:
+
+**On control plane node 1:**
+
+```bash
+kubeadm init --upload-certs
+```
+
+The new `--upload-certs` flag will trigger a new kubeadm init phase executing following
+actions:
+
+1. An encryption key will be generated; such **key will never be stored on cluster**
```

Comment:
what is the key length? encryption method?

Reply:
my proposal is 32bytes for key (SHA-256) with AES-256/GCM as the method.
we might want to include the details on that, but others disagree:
https://github.com/kubernetes/enhancements/pull/713#discussion_r249958845

Reply:
also by using a 32byte cryto-secure checksum (e.g. SHA-256) as key to be fed to AES, we can technically use a shorter key on the user side. 
as long as SHA-256 is not a security risk this method can slightly simplify the command line.

Reply:
> my proposal is 32bytes for key (SHA-256) with AES-256/GCM as the method

+1

> by using a 32byte cryto-secure checksum (e.g. SHA-256) as key to be fed to AES, we can technically use a shorter key on the user side.

if the input to the hash (length and charset) is known, using a hash to expand the data size does not increase entropy, so is not a good idea. to take an extreme example, if the key had a max length of 1 and possible characters 0-9, you could brute force by trying just 10 sha256 hashes.

Reply:
that is true, the extra step will reduce the entropy.

i guess i meant to simplify the command line - e.g.:
`abcdef.1234567890abcdef` (of which we then take the checksum)
vs
`37aa9a83376e16819b4d826732dc0d8a8a310c1b9665f0082ac35e3f04cdd65d`

but we should probably just pass the whole 32bytes.